### PR TITLE
Common Container struct

### PR DESCRIPTION
**Pull request by janslow**
Extract a `Container` struct (and `ContainerSource`) which represents the common properties of `Sidecar`, `InitContainer`, `Worker`, `Job` and `CronJob`. This also standardises the properties of them all (e.g., they all have `env` and `preserveEnv` properties, fixing #167).

I've also removed the `preserveEnv` property from `shipcat_definitions`, the behaviour of this is now handled in `shipcat_filebacked`, by merging the Worker's (etc) env with the parent when building the main structs (so it no longer requires custom logic in the Helm templates).

Also introduce a `Build` trait for the `*Source` structs, to make it simpler to build sources into the main structs (it may be possible to derive this in the future).

There shouldn't be any breaking changes in this (`preserveEnv` is no longer included in the Helm values, but that's equivalent to `preserveEnv: false`).

### CODE REVIEW

**Comment by janslow** [comment]
I think this is code complete, so ready for code review, but I want to do a full diff of the Helm/CRD outputs and fix gradual rollouts (in another PR) first.

**Comment by janslow** [comment]
I've diffed against `dev-uk` and `prod-uk` (`shipcat kong`, `shipcat helm $svc template` and `shipcat crd $svc`), there are no major changes.

The changes to the helm output are either whitespace or slight changes to env vars. The latter is for services which were using `preserveEnv` but overriding the value (originally, this wrote two env var entries, now it only writes the override).

The only change to the CRDs (after running it through `yq --sort-keys`) are the removal of the `preserveEnv` attribute, but `raftcat` doesn't use that.

File: shipcat_definitions/src/structs/resources.rs [hunk 1]
```diff
@@ -21,37 +20,26 @@ pub struct ResourceRequest<T> {
     // TODO: ephemeral-storage + extended-resources
 }
 
-/// Kubernetes resource limits
-#[derive(Serialize, Deserialize, Clone)]
-#[cfg_attr(feature = "filesystem", serde(deny_unknown_fields))]
-pub struct ResourceLimit<T> {
```
[the remark this answers is not in the document]
Comment (clux) [reply shipcat_definitions/src/structs/resources.rs]: That sounds good. Thanks for the link.

`type ResourceList map[ResourceName]resource.Quantity` <- classic.

[the remark this answers is not in the document]
Comment (janslow) [reply shipcat_definitions/src/structs/resources.rs]: Ok, I'll do that in a separate PR, as it's going to create a big diff unrelated to most of the other changes.

[the remark this answers is not in the document]
Comment (clux) [reply shipcat_definitions/src/structs/resources.rs]: That sounds entirely sensible.

I love how they've literally aliased a list to a map..

```go
type ResourceList map[ResourceName]resource.Quantity
```

File: shipcat_filebacked/src/util/build.rs [hunk 2]
```diff
@@ -6,13 +6,7 @@ pub trait Build<T, P> {
...
 impl<T, P, S: Build<T, P>> Build<Option<T>, P> for Option<S> {
     fn build(self, params: &P) -> Result<Option<T>> {
-        match self {
-            None => Ok(None),
-            Some(s) => {
-                let t = s.build(params)?;
-                Ok(Some(t))
-            }
-        }
+        self.map(|s| s.build(params)).transpose()
```
Comment (clux) [inline shipcat_filebacked/src/util/build.rs]: :clap:

**Pull request merged**